File: src/libCamera/src/Camera.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <math.h>
#include <Eigen/Geometry>
#include <Camera/Camera.h>

const double DEG_TO_RAD = M_PI / 180.0;

Camera::Camera(const float position[3],
               const float view[3], const float up[3], const float res[2],
               const float fov[2], float f) : camera_origin{position[0], position[1], position[2]},
                                              looking_at{view[0], view[1], view[2]},
                                              up{up[0], up[1], up[2]},
                                              resolution{res[0], res[1]},
                                              field_of_view{fov[0] * DEG_TO_RAD, fov[1] * DEG_TO_RAD},
                                              focal_length{f} {
    compute_camera_parms();
}

void
Camera::set_image_size( unsigned int width, unsigned int height ) {
    // Compute pixel dimensions in world units
    resolution.x() = width;
    resolution.y() = height;
    pixel_width = image_plane_dimensions.x() / resolution.x();
    pixel_height = image_plane_dimensions.y() / resolution.y();
}

void
Camera::compute_camera_parms() {
    using namespace Eigen;

    // n is normal to image plane, points in opposite direction of view point
    Vector3f N{camera_origin.x() - looking_at.x(),
               camera_origin.y() - looking_at.y(),
               camera_origin.z() - looking_at.z()};
    n = N.normalized();

    // u is a vector that is perpendicular to the plane spanned by
    // N and view up vector (cam->up), ie in the image plane and horizontal
    Vector3f U = up.cross(n);
    u = U.normalized();

    // v is a vector perpendicular to N and U, i.e vertical in image palne
    v = n.cross(u);

    // Compute dimensions of image plane in world units
    /*
     * We have
     *
     * +          ^
     * |\         |
     * |a\        |
     * |  \       f = F*pixels per world unit
     * |   \      |
     * +----+     v
     * pix width /w
     * tan(a) = opp/adj = (image_width_pixels * 0.5) / focal_length_pixels
     * 2tan(a) = image_width_pixels / focal_length_pixels
     * => focal_length_pixels = image_width_pixels / (2tan(a))
     *
     * pixels_per_world_unit = focal_length_pixels / focal_length_world_units
     *  focal_length_pixels = pixels_per_world_unit / focal_length_world_units
     *
     *  image_width_pixels / (2tan(a)) = pixels_per_world_unit / focal_length_world_units
     *  => pixels_per_world_unit = (image_width_pixels / (2tan(a))) *  focal_length_world_units
     */
    double pixels_per_wux = (resolution.x() / (2 * tan(field_of_view.x() * 0.5f))) * focal_length;
    double pixels_per_wuy = (resolution.y() / (2 * tan(field_of_view.y() * 0.5f))) * focal_length;

    double image_plane_height = pixels_per_wuy / resolution.y();
    double image_plane_width = pixels_per_wux / resolution.x();

    Vector3f image_plane_centre = camera_origin - (n * focal_length);
    image_plane_origin = image_plane_centre - (u * image_plane_width * 0.5f) - (v * image_plane_height * 0.5f);

    // Compute pixel dimensions in world units
    pixel_width = image_plane_width / resolution.x();
    pixel_height = image_plane_height / resolution.y();

    image_plane_dimensions.x() = image_plane_width;
    image_plane_dimensions.y() = image_plane_height;
}


void parseVec3(const std::string &str, float vec[3]) {
    using namespace std;

    istringstream pos(str);

    string val;
    getline(pos, val, ',');
    float x = stof(val);
    getline(pos, val, ',');
    float y = stof(val);
    getline(pos, val);
    float z = stof(val);
    vec[0] = x;
    vec[1] = y;
    vec[2] = z;
}

void parseVec2(const std::string &str, float vec[2]) {
    using namespace std;

    istringstream pos(str);

    string val;
    getline(pos, val, ',');
    float x = stof(val);
    getline(pos, val, ',');
    float y = stof(val);
    vec[0] = x;
    vec[1] = y;
}

Camera loadCameraFromFile(const std::string &filename) {
    using namespace std;
    using namespace Eigen;

    bool fl_position, fl_view, fl_up, fl_resolution, fl_fov, fl_f;
    fl_position = fl_view = fl_up = fl_resolution = fl_fov = fl_f = false;

    ifstream file;
    file.exceptions(ifstream::failbit | ifstream::badbit);

    try {
        file.open(filename);
    }
    catch (ifstream::failure& e) {
        cout << "ERROR::CAMFILE::NOT_FOUND " << filename << endl;
        cerr << strerror(errno) << endl;
        throw runtime_error("No camera file found" );
    }

    float position[3];
    float view_direction[3];
    float up_direction[3];
    float resolution[2];
    float field_of_view[2];
    float focal_length;

    file.exceptions(ifstream::goodbit);
    string line;
    while (getline(file, line)) {
        istringstream is_line(line);
        string key;
        if (getline(is_line, key, '=')) {
            string value;
            if (getline(is_line, value)) {
                // Handle the line
                if (key == "position") {
                    parseVec3(value, position);
                    fl_position = true;
                } else if (key == "view") {
                    parseVec3(value, view_direction);
                    fl_view = true;
                } else if (key == "up") {
                    parseVec3(value, up_direction);
                    fl_up = true;
                } else if (key == "resolution") {
                    parseVec2(value, resolution);
                    fl_resolution = true;
                } else if (key == "fov") {
                    parseVec2(value, field_of_view);
                    fl_fov = true;
                } else if (key == "f") {
                    focal_length = stof(value);
                    fl_f = true;
                } else {
                    cerr << "ERROR::CAMFILE::UNKNOWN_KEY " << key << endl;
                    throw std::domain_error("CAMFILE::UNKNOWN_KEY");
                }
            }
        }
    }

    if (!(fl_position && fl_view && fl_up && fl_resolution && fl_fov && fl_f)) {
        cerr << "ERROR::CAMFILE::MISSING_KEY" << endl;
        throw std::domain_error("CAMFILE::MISSING_KEY");
    }

    return Camera{position, view_direction, up_direction, resolution, field_of_view, focal_length};
}

/*
 * Get the camera matrix.
 */
void Camera::camera_intrinsics(Eigen::Matrix3f &K) {
    double cx = resolution.x() * 0.5;
    double cy = resolution.y() * 0.5;
    double fx = resolution.x() / tan(field_of_view.x() * 0.5);
    double fy = resolution.y() / tan(field_of_view.y() * 0.5);
    float skew = 0.0f;
    K << fx, skew, cx,
            0.0f, fy, cy,
            0.0f, 0.0f, 1.0f;
}

/**
 * Based on https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/gluLookAt.xml
 */
void
Camera::camera_extrinsics(Eigen::Matrix3f &R, Eigen::Vector3f &t) {
    using namespace Eigen;

    R << v.x(), u.x(), n.x(),
            v.y(), u.y(), n.y(),
            v.z(), u.z(), n.z();

    t = -R * camera_origin;
}

/**
 * Break camera down i to extrinsic and intrinsic matrices
 */
void
Camera::decompose(Eigen::Matrix3f &K, Eigen::Matrix3f &R, Eigen::Vector3f &t) {
    camera_intrinsics(K);
    camera_extrinsics(R, t);
    std::cout << "Cam R " << R << std::endl;
    std::cout << "Cam t " << t << std::endl;
}

/**
 * Friend function for dumping Camera objects
 */
std::ostream
&operator<<(std::ostream &os, const Camera &camera) {
    using namespace std;
    os << "pos : " << camera.camera_origin.x() << ", " << camera.camera_origin.y() << ", " << camera.camera_origin.z() << endl;
    os << "vew : " << camera.n.x() << ", " << camera.n.y() << ", " << camera.n.z() << endl;
    os << " up : " << camera.v.x() << ", " << camera.v.y() << ", " << camera.v.z() << endl;
    os << "res : " << camera.resolution.x() << ", " << camera.resolution.y() << endl;
    os << "fov : " << camera.field_of_view.x() << ", " << camera.field_of_view.y() << endl;
    os << "foc : " << camera.focal_length << endl;
    return os;
}

/*
 * Compute the backprojection of a point from X,Y and depth in world space
 */
Eigen::Vector3f
Camera::to_world_coordinates(unsigned int pixel_x, unsigned int pixel_y, float depth) const {
    using namespace Eigen;

    // Get world coordinates of pixel through which the ray passes
   Vector3f pixelCoordinate = image_plane_origin
                               + ((pixel_x /*+ 0.5*/) * pixel_width * u)
                               + ((pixel_y /*+ 0.5*/) * pixel_height * v);

   // Now project the ray to the depth expected to give the final world coordinate of the projected point
    Vector3f rayDirection = (pixelCoordinate - camera_origin).normalized();
    return camera_origin + (rayDirection * depth);
}

/*
 * Convenience method to to_world_coords a point into an array
 */
void
Camera::to_world_coordinates(unsigned int pixel_x, unsigned int pixel_y, float depth, float *world_coordinate) const {
    assert( world_coordinate != nullptr);
    Eigen::Vector3f point = to_world_coordinates(pixel_x, pixel_y, depth);
    world_coordinate[0] = point(0);
    world_coordinate[1] = point(1);
    world_coordinate[2] = point(2);
}

/**
 * Change the location of the camera in world coordinates.
 * If keep_facing is true, the camera stays pointed in the same direction as it currently is
 * otherwise, it rotates to stay pointing at the same point in space.
 */
void
Camera::move_to( float world_x, float world_y, float world_z, bool keep_facing ) {
    camera_origin << world_x, world_y, world_z;
    if( keep_facing) {
        looking_at << world_x - n.x(), world_y - n.y(), world_z - n.z();
    }
    compute_camera_parms();
}

/**
 * Change the point that the camera is looking at
 */
void
Camera::look_at( float view_x, float view_y, float view_z ) {
    looking_at << view_x, view_y, view_z;
    compute_camera_parms();
}

File: java/animesh-tools/animeshtools/src/main/java/org/ddurbin/common/Camera.java
package org.ddurbin.common;

import java.io.*;

public class Camera {
    private final Vector3f position;
    private final Vector3f view;
    private final Vector3f up;
    private final Vector2f resolution;
    private final Vector2f fov;
    private final float focalDistance;
    private Matrix3f rotation;
    private Matrix3f k;
    private Vector3f translation;
    private CamCoordSystem camCoordinateSystem;

    public Camera(Vector3f position, Vector3f view, Vector3f up, Vector2f resolution, Vector2f fov, float focalDistance) {
        this.position = position;
        this.view = view;
        this.up = up;
        this.resolution = resolution;
        this.fov = fov;
        this.focalDistance = focalDistance;
    }

    /**
     * Load camera settings from a file.
     */
    public static Camera loadFromFile(String filename) throws IOException {
        boolean flPosition, flView, flUp, flResolution, flFov, flF;
        flPosition = flView = flUp = flResolution = flFov = flF = false;

        Vector3f position = null, view = null, up = null;
        Vector2f resolution = null, fov = null;
        float focalDistance = 0.0f;

        LineNumberReader lnr = new LineNumberReader(new FileReader(new File(filename)));

        String line;
        while ((line = lnr.readLine()) != null) {
            String[] keyValue = line.split("=");
            if (keyValue.length != 2) {
                throw new IllegalArgumentException("Bad line in camera file : " + line);
            }
            String key = keyValue[0].trim();
            String value =keyValue[1].trim();
            switch (key) {
                case "position":
                    position = Mapper.mapVector3f (value);
                    flPosition = true;
                    break;
                case "view":
                    view = Mapper.mapVector3f(value);
                    flView = true;
                    break;
                case "up":
                    up = Mapper.mapVector3f (value);
                    flUp = true;
                    break;
                case "resolution":
                    resolution = Mapper.mapVector2f (value);
                    flResolution = true;
                    break;
                case "fov":
                    fov = Mapper.mapVector2f (value);
                    flFov = true;
                    break;
                case "f":
                    focalDistance = Float.valueOf(value);
                    flF = true;
                    break;
                default:
                    throw new IllegalArgumentException("CAMFILE::UNKNOWN_KEY " + key);
            }
        }
        if (!(flPosition && flView && flUp && flResolution && flFov && flF)) {
            throw new IllegalArgumentException("CAMFILE::MISSING_KEY");
        }

        return new Camera(position, view, up, resolution, fov, focalDistance);
    }

    /*
     * Get the camera matrix
     */
    Matrix3f getIntrinsicMatrix() {
        if (k == null) {
            float cx = resolution.x / 2.0f;
            float cy = resolution.y / 2.0f;
            float fx = (float) (resolution.x / Math.tan(fov.x / 2.0f));
            float fy = (float) (resolution.y / Math.tan(fov.y / 2.0f));
            float skew = 0.0f;

            k = new Matrix3f(fx, skew, cx, 0.0f, fy, cy, 0.0f, 0.0f, 1.0f);
        }
        return k;
    }

    /**
     * Based on https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/gluLookAt.xml
     */
    public Matrix3f getRotationMatrix() {
        if (rotation == null) {
            Vector3f forward = view.minus(position);
            forward = forward.normalized();

            Vector3f upx = up.normalized();
            Vector3f s = forward.cross(upx);
            s = s.normalized();

            Vector3f u = s.cross(forward);

            rotation = new Matrix3f(s.x, u.x, -forward.x,
                    s.y, u.y, -forward.y,
                    s.z, u.z, -forward.z);
        }
        return rotation;
    }

    public Vector3f getTranslationVector() {
        if (translation == null) {
            translation = getRotationMatrix().times(position).times(-1);
        }
        return translation;
    }

    private class CamCoordSystem {
        Vector3f origin;
        Vector3f n;
        Vector3f u;
        Vector3f v;
        Vector3f imagePlaneOrigin;
        Vector2f imagePlaneDimensions;

    }

    /**
     * Construct an eye coordinate system
     * Input: camera position, center of interest, view-up vector
     * Returns: new origin and three basis vectors
     * <p>
     *      /|
     *     / |
     *    /  |
     *   /   |
     *  / ^  |
     * / v|  |
     * |  |-----> n
     * | /  /
     * |Lu /
     * |  /
     * | /
     * |/
     */
    private CamCoordSystem constructCameraCoordinateSystem() {
        CamCoordSystem ccs = new CamCoordSystem();

        // n is normal to image plane, points in opposite direction of view point
        ccs.n = new Vector3f(position.x - view.x, position.y - view.y, position.z - view.z).normalized();

        // u is a vector that is perpendicular to the plane spanned by
        // N and view up vector (cam->up)
        ccs.u = up.cross(ccs.n).normalized();

        // v is a vector perpendicular to N and U
        ccs.v = ccs.n.cross(ccs.u);

        // origin is cam-centre
        ccs.origin = position;
        return ccs;
    }


    private void constructImagePlaneOrigin(CamCoordSystem ccs) {
        float imagePlaneHeight = (float) (Math.tan(Math.toRadians(fov.y * 0.5f)) * 2.0f * focalDistance);
        float imagePlaneWidth = (float) (Math.tan(Math.toRadians(fov.x * 0.5f)) * 2.0f * focalDistance);

        Vector3f imagePlaneCentre = position.minus((ccs.n.times(focalDistance)));

        ccs.imagePlaneOrigin = imagePlaneCentre.minus((ccs.u.times(imagePlaneWidth * 0.5f)).minus(ccs.v.times(imagePlaneHeight * 0.5f)));
        ccs.imagePlaneDimensions = new Vector2f(imagePlaneWidth, imagePlaneHeight);
    }

    /*
     * Compute the backprojection of a point from X,Y and depth plus camera
     */
    public Vector3f to_world_coordinates(int pixelX, int pixelY, float depth) {
        if( camCoordinateSystem == null ) {
            camCoordinateSystem  = constructCameraCoordinateSystem();
            constructImagePlaneOrigin(camCoordinateSystem);
        }

        // Compute pixel coordinate in world space
        float pixelWidth = camCoordinateSystem.imagePlaneDimensions.x / resolution.x;
        float pixelHeight = camCoordinateSystem.imagePlaneDimensions.y / resolution.y;
        Vector3f pixelCoordinate = camCoordinateSystem.imagePlaneOrigin
                .plus(camCoordinateSystem.u.times(pixelX * pixelWidth))
                .plus(camCoordinateSystem.v.times(pixelY * pixelHeight));

        Vector3f rayDirection = (pixelCoordinate.minus(camCoordinateSystem.origin)).normalized();
        return camCoordinateSystem.origin.plus(rayDirection.times(depth));
    }

    public String toString() {
        return String.format("pos : (%.3f, %.3f, %.3f)\n vew : (%.3f, %.3f, %.3f)\n up : (%.3f, %.3f, %.3f)\nres : (%.3f, %.3f)\nfov : (%.3f, %.3f)\nfoc : %f\n",
                position.x, position.y, position.z,
                view.x, view.y, view.z,
                up.x, up.y, up.z,
                resolution.x, resolution.y,
                fov.x, fov.y, focalDistance);
    }
}
